### PR TITLE
Enable users to select models by label 

### PR DESCRIPTION
**Pull request by danielholanda**
Closes https://github.com/groq/mlagility/issues/140

This MR enables users to filter models by label

### CODE REVIEW

File: docs/tools_user_guide.md [hunk 1]
```diff
@@ -322,6 +335,11 @@ You can also leverage model hashes (see [Model Hashes](#model-hashes)) to filter
 
 > See the [Filtering Model Hashes tutorial](https://github.com/groq/mlagility/blob/main/examples/cli/discovery.md#filtering-model-hashes) for a detailed example.
 
+Additionally, you can leverage labels (see [Labels](#labels)) to filter which models in a script will be acted on, in the following manner:
+  - `benchit *.py --label test_group::a` will only benchmark the scripts lavels with `test_group::a`.
```
Comment (jeremyfowers) [inline docs/tools_user_guide.md:339]:
```suggestion
  - `benchit *.py --label test_group::a` will only benchmark the scripts labels with `test_group::a`.
```

File: docs/tools_user_guide.md [hunk 2]
```diff
@@ -322,6 +335,11 @@ You can also leverage model hashes (see [Model Hashes](#model-hashes)) to filter
...
+  - You can also supply multiple labels, for example `benchit *.py --label test_group::a domain::nlp` only benchmark models that have both `test_group::a`, and `domain::nlp` labels.
```
Comment (jeremyfowers) [inline docs/tools_user_guide.md:340]:
```suggestion
  - You can also supply multiple labels, for example `benchit *.py --label test_group::a domain::nlp` only benchmark scripts that have both `test_group::a`, and `domain::nlp` labels.
```

File: examples/cli/discovery.md [hunk 3]
```diff
@@ -220,6 +220,49 @@ You can see that both models are discovered, but only `another_pytorch_model` wa
 
 > See the [Input Script documentation](https://github.com/groq/mlagility/blob/main/docs/tools_user_guide.md#input-script) for more details.
 
+## Filtering Model Labels
+
+You can also leverage the labels feature of `benchit` to filter which models are acted on. Labels are pragmas added by the user to the first line of a `.py` file to list some of the attributes of that given script. `two_models.py`, for example has the label `test_group::b`, while `hello_world.py` and `max_depth.py` have the label `test_group::b`.
```
Comment (jeremyfowers) [inline examples/cli/discovery.md:225]:
```suggestion
You can also leverage the labels feature of `benchit` to filter which models are acted on. Labels are pragmas added by the user to the first line of a `.py` file to list some of the attributes of that given script. `two_models.py`, for example has the label `test_group::b`, while `hello_world.py` and `max_depth.py` have the label `test_group::a`.
```

File: examples/cli/discovery.md [hunk 4]
```diff
@@ -220,6 +220,49 @@ You can see that both models are discovered, but only `another_pytorch_model` wa
...
+two_models.py:
+        another_pytorch_model (executed 1x)
+                Model Type:     Pytorch (torch.nn.Module)
+                Class:          SmallModel (<class 'two_models.SmallModel'>)
+                Location:       /home/jfowers/mlagility/examples/cli/scripts/two_models.py, line 40
+                Parameters:     510 (<0.1 MB)
+                Hash:           215ca1e3
+                Status:         Model successfully benchmarked on Intel(R) Xeon(R) CPU @ 2.20GHz
+                                Mean Latency:   0.000   milliseconds (ms)
+                                Throughput:     509528.6        inferences per second (IPS)
```
Comment (jeremyfowers) [inline examples/cli/discovery.md:248]: `two_models.py` has two models, why is only one showing up here?

File: examples/cli/discovery.md [hunk 5]
```diff
@@ -220,6 +220,49 @@ You can see that both models are discovered, but only `another_pytorch_model` wa
...
+max_depth.py:
+        pytorch_model (executed 1x)
+                Model Type:     Pytorch (torch.nn.Module)
+                Class:          TwoLayerModel (<class 'max_depth.TwoLayerModel'>)
+                Location:       /home/jfowers/mlagility/examples/cli/scripts/max_depth.py, line 41
+                Parameters:     85 (<0.1 MB)
+                Hash:           80b93950
+                Status:         Model successfully benchmarked on Intel(R) Xeon(R) CPU @ 2.20GHz
+                                Mean Latency:   0.000   milliseconds (ms)
+                                Throughput:     693955.3        inferences per second (IPS)
```
Comment (jeremyfowers) [inline examples/cli/discovery.md:259]: `max_depth.py` has the label `test_group::a`, why is it showing up here?

File: src/mlagility/api/script_api.py [hunk 6]
```diff
@@ -8,6 +8,7 @@
 from groqflow.justgroqit.stage import Sequence
 import mlagility.cli.slurm as slurm
 import mlagility.common.filesystem as filesystem
+import mlagility.common.labels as lab
```
Comment (jeremyfowers) [inline src/mlagility/api/script_api.py:11]:
```suggestion
import mlagility.common.labels as labels_library
```

Style: don't use uncommon abbreviations. If you're unable to use `labels` due to a namespace collision, `labels_library` is a reasonable name.

File: src/mlagility/api/script_api.py [hunk 7]
```diff
@@ -146,8 +148,27 @@ def benchmark_script(
     models_found: Dict[str, ModelInfo] = {}
 
     for script in input_scripts:
+        script_path, targets, encoded_input = decode_input_script(script)
+
+        # Filter scripts by label
+        if labels != []:
```
Comment (jeremyfowers) [inline src/mlagility/api/script_api.py:154]: This is a spicy test because the `labels` arg defaults to `None` in this function. Should it be `if labels is not None and labels != []`?

Or if you want to play fast and loose I believe you can just do `if labels`, since that will filter out both `None` and `[]`.

File: src/mlagility/api/script_api.py [hunk 8]
```diff
@@ -146,8 +148,27 @@ def benchmark_script(
...
+                if key not in file_labels:
+                    skip_script = True
+                    break
+                # A label key may point to multiple label values
+                # Skip if not all values of user_labels[key] are in file_labels[key]
+                elif not all(elem in user_labels[key] for elem in file_labels[key]):
+                    skip_script = True
+                    break
+            if skip_script:
+                continue
```
Comment (jeremyfowers) [inline src/mlagility/api/script_api.py:169]:
```suggestion
            user_labels = lab.to_dict(labels)
            file_labels = lab.load_from_file(encoded_input)
            for key in user_labels:
                # Skip benchmarking if user_label is not in file_labels, or...
                # A label key may point to multiple label values
                # Skip if not all values of user_labels[key] are in file_labels[key]
                if key not in file_labels or not all(elem in user_labels[key] for elem in file_labels[key]):
                    continue
```

Can be a lot shorter 😸 

There's no need to `break` to help get to a `continue` statement, just `continue`.

Comment (jeremyfowers) [inline src/mlagility/api/script_api.py:169]: This is pretty verbose because you're using a `continue` statement within a nested loop. 

Suggest either:
1. Move this logic outside of the `for script in input_scripts:` loop so that you can simply `remove()` unwanted scripts from `input_scripts` (in which case the filtering in #156 should do the same)
2. Rewrite this code block to remove the inner loop

File: src/mlagility/api/script_api.py [hunk 9]
```diff
@@ -146,8 +148,27 @@ def benchmark_script(
     models_found: Dict[str, ModelInfo] = {}
 
     for script in input_scripts:
+        script_path, targets, encoded_input = decode_input_script(script)
+
+        # Filter scripts by label
+        if labels != []:
+            skip_script = False
+            user_labels = lab.to_dict(labels)
+            file_labels = lab.load_from_file(encoded_input)
```
Comment (jeremyfowers) [inline src/mlagility/api/script_api.py:157]: The naming here gives me a false equivalency between `user_labels` and `file_labels`. Yes, they are both `labels`, but they are not both used to label things (ie, the user put some labels, and the file put some labels). IMO `user_labels` is really more like `required_labels`.

Also suggest renaming `file_labels` to `script_labels`, since we refer to it as `script` everywhere else.

File: src/mlagility/api/script_api.py [hunk 10]
```diff
@@ -146,8 +148,27 @@ def benchmark_script(
...
+            for key in user_labels:
```
Comment (jeremyfowers) [inline src/mlagility/api/script_api.py:158]: nit: `key` makes me think of a `dict`, but there is no `dict` here. Why not just `label` or `required_label`?

File: src/mlagility/api/script_api.py [hunk 11]
```diff
@@ -146,8 +148,27 @@ def benchmark_script(
...
+                # Skip benchmarking if user_label is not in file_labels
+                if key not in file_labels:
+                    skip_script = True
+                    break
+                # A label key may point to multiple label values
+                # Skip if not all values of user_labels[key] are in file_labels[key]
+                elif not all(elem in user_labels[key] for elem in file_labels[key]):
```
Comment (jeremyfowers) [inline src/mlagility/api/script_api.py:165]: Isn't this requirement a superset of the `if key not in file_labels:` requirement? (can't you just delete `if key not in file_labels:`?)

Comment (jeremyfowers) [reply src/mlagility/api/script_api.py:165]: Or maybe I'm just confused... 

This statement is using `user_labels[key]` but it's in a loop that's doing `for key in user_labels:`


that could at least be `for key, value in user_labels.items()` and then `all(elem in value...` right?

File: src/mlagility/cli/cli.py [hunk 12]
```diff
@@ -130,6 +131,14 @@ def main():
         default=filesystem.DEFAULT_CACHE_DIR,
     )
 
+    benchmark_parser.add_argument(
+        "--label",
+        dest="labels",
+        help="Only benchmark the models that have the provided labels",
```
Comment (jeremyfowers) [inline src/mlagility/cli/cli.py:137]:
```suggestion
        help="Only benchmark the scripts that have the provided labels",
```

File: test/cli.py [hunk 13]
```diff
@@ -533,6 +533,54 @@ def test_011_cli_benchmark(self):
...
+            "--label",
+            "test_group::a",
+            "--build-only",
+            "--cache-dir",
+            cache_dir,
+        ]
+        with patch.object(sys, "argv", flatten(testargs)):
+            benchitcli()
+
+        assert len(cache.get_all(cache_dir)) == 1
```
Comment (jeremyfowers) [inline test/cli.py:552]: I'm nervous that we're adding more tests that make assumptions about the contents of the test corpus, which is shared amongst all tests. For example, if I were to use linear.py as a starting point for a new input script for the corpus, then that new script would also have `test_group::a` at the top of it. 

Is there a way to code this test so that it doesn't have a baked in assumption about the number of scripts with `test_group::a`?

**Review by jeremyfowers — CHANGES_REQUESTED** [review]
Overall its a very slick feature from a UX perspective, but I am concerned about maintainability. Please take a look at some of the implementation details I have pointed out in the comments.

File: src/mlagility/api/script_api.py (hunk 8 above)
Comment (jeremyfowers) [reply src/mlagility/api/script_api.py:169]: Since #156 just merged, please make sure the filtering method used here is stylistically consistent with the `--resume` filtering (feel free to refactor the `--resume` filtering if needed).

File: examples/cli/discovery.md (hunk 5 above)
Comment (danielholanda) [reply examples/cli/discovery.md]: Addressed that!

File: src/mlagility/api/script_api.py (hunk 6 above)
Comment (danielholanda) [reply src/mlagility/api/script_api.py:11]: Accepted suggestion.

File: src/mlagility/api/script_api.py (hunk 7 above)
Comment (danielholanda) [reply src/mlagility/api/script_api.py:154]: Accepted suggestion

File: src/mlagility/api/script_api.py (hunk 9 above)
Comment (danielholanda) [reply src/mlagility/api/script_api.py:157]: Changed names according to suggestion

File: examples/cli/discovery.md (hunk 4 above)
Comment (danielholanda) [reply examples/cli/discovery.md]: Fixed!

File: src/mlagility/api/script_api.py (hunk 11 above)
Comment (danielholanda) [reply src/mlagility/api/script_api.py]: I think you are a bit confused.

`[elem in required_labels[key] for elem in script_labels[key]]` is inside a loop that is iterating over all keys of `required_labels`. If a key of `required_labels` doesn't exist in `script_labels` that will result in a `KeyError` exception when evaluating `script_labels[key]`. 

This is why checking `if key not in script_labels` is needed.

I slightly modified the comments to clarify that.

File: src/mlagility/api/script_api.py (hunk 8 above)
Comment (danielholanda) [reply src/mlagility/api/script_api.py]: I'm now checking whether `required_labels` is a subset of `script_labels` in a function, which makes this code a lot less verbose. Let me know what you think.

File: test/cli.py (hunk 13 above)
Comment (danielholanda) [reply test/cli.py]: Modified to ensure that we will not break this test if we accidentally add another test with an existing `test_group` label. 

Let me know what you think. 

We can also revert that change (to be less verbose) and simply do not add the `test_group` label to other test scripts in the future.

File: src/mlagility/cli/cli.py [hunk 14]
```diff
@@ -131,6 +132,14 @@ def main():
         default=filesystem.DEFAULT_CACHE_DIR,
     )
 
+    benchmark_parser.add_argument(
+        "--label",
```
Comment (jeremyfowers) [inline src/mlagility/cli/cli.py:136]: Should this be `--labels` since its a list? the `dest` is even `labels`.

Fix in the docs as well if you decide to change it.

Comment (jeremyfowers) [reply src/mlagility/cli/cli.py:136]: My vote is for it to be `--labels`, especially since the `benchmark_scripts()` API argument is `labels`

**Review by jeremyfowers — APPROVED** [review]
Approved so you can merge after https://github.com/groq/mlagility/pull/152/files#r1119335135 is fixed

File: src/mlagility/api/script_api.py (hunk 11 above)
Comment (kovtcharov) [reply src/mlagility/api/script_api.py]: FYI, you can use `script_labels.get(key)` if you want to avoid `KeyError`.

File: docs/tools_user_guide.md [hunk 15]
```diff
@@ -322,6 +335,11 @@ You can also leverage model hashes (see [Model Hashes](#model-hashes)) to filter
 
 > See the [Filtering Model Hashes tutorial](https://github.com/groq/mlagility/blob/main/examples/cli/discovery.md#filtering-model-hashes) for a detailed example.
 
+Additionally, you can leverage labels (see [Labels](#labels)) to filter which models in a script will be acted on, in the following manner:
+  - `benchit *.py --label test_group::a` will only benchmark the scripts labels with `test_group::a`.
```
Comment (kovtcharov) [inline docs/tools_user_guide.md:339]: Perhaps consider calling it `--labels` instead since it supports multiple labels?

File: docs/tools_user_guide.md [hunk 16]
```diff
@@ -322,6 +335,11 @@ You can also leverage model hashes (see [Model Hashes](#model-hashes)) to filter
...
+  - You can also supply multiple labels, for example `benchit *.py --label test_group::a domain::nlp` only benchmark scripts that have both `test_group::a`, and `domain::nlp` labels.
```
Comment (kovtcharov) [inline docs/tools_user_guide.md:340]: What happens if a script does not contain the pragmas? How does `benchit` handle it?

File: src/mlagility/cli/cli.py (hunk 14 above)
Comment (kovtcharov) [reply src/mlagility/cli/cli.py:136]: Made the same comment above before seeing this, vote +1.

**Review by kovtcharov — APPROVED** [review]

File: docs/tools_user_guide.md (hunk 15 above)
Comment (danielholanda) [reply docs/tools_user_guide.md:339]: Switched to `--labels`

File: docs/tools_user_guide.md (hunk 16 above)
Comment (danielholanda) [reply docs/tools_user_guide.md:340]: Yes. Pragmas are optional. Modified the documentation to make that clear
